graphql_compiler/schema/schema_info.py

```diff
@@ -1,13 +1,116 @@
 # Copyright 2019-present Kensho Technologies, LLC.
 from collections import namedtuple
+from dataclasses import dataclass
+from functools import partial
+from typing import Dict, Optional
 
+from graphql.type import GraphQLSchema
 from graphql.type.definition import GraphQLInterfaceType, GraphQLObjectType
 import six
 import sqlalchemy
+from sqlalchemy.dialects.postgresql import dialect as postgresql_dialect
+from sqlalchemy.engine.interfaces import Dialect
 
 from . import is_vertex_field_name
 
 
+# Describes the intent to join two tables using the specified columns.
+#
+# The resulting join expression could be something like:
+# JOIN origin_table.from_column = destination_table.to_column
+#
+# The type of join (inner vs left, etc.) is not specified.
+# The tables are not specified.
+DirectJoinDescriptor = namedtuple(
+    "DirectJoinDescriptor",
+    (
+        "from_column",  # The column in the source table we intend to join on.
+        "to_column",  # The column in the destination table we intend to join on.
+    ),
+)
+
+
+@dataclass
+class SchemaInfo:
+    """Class to store all schema info required for querying."""
+
+    # This is a required component of any SchemaInfo objects.
+    schema: GraphQLSchema
+
+    # Optional dict of GraphQL interface or type -> GraphQL union.
+    # Used as a workaround for GraphQL's lack of support for
+    # inheritance across "types" (i.e. non-interfaces), as well as a
+    # workaround for Gremlin's total lack of inheritance-awareness.
+    # The key-value pairs in the dict specify that the "key" type
+    # is equivalent to the "value" type, i.e. that the GraphQL type or
+    # interface in the key is the most-derived common supertype
+    # of every GraphQL type in the "value" GraphQL union.
+    # Recursive expansion of type equivalence hints is not performed,
+    # and only type-level correctness of this argument is enforced.
+    # See README.md for more details on everything this parameter does.
+    # *****
+    # Be very careful with this option, as bad input here will
+    # lead to incorrect output queries being generated.
+    # *****
+    type_equivalence_hints: Dict[str, str]  # Optional
+
+    # ####
+    # The fields below are required for SQL databases.
```

Comment:
I'm not a huge fan of this layout, where these fields are required in one case (SQL) and disallowed (ignored and must not be set) in another case (non-SQL). I'm worried this could end up causing additional complexity and allow bugs to creep in. For example, we'll have to validate these fields differently depending on the backend, which means that having constructed `SchemaInfo` instance does not necessarily mean that the instance is valid and usable.

Instead, consider a `SchemaInfo[BackendSpecificDataT]` generic dataclass, where the `BackendSpecificDataT` can contain different things for different backends. We may then be able to use `NewType` from [typing](https://docs.python.org/3/library/typing.html#newtype) to create "new" types for each of the few combinations of `SchemaInfo` and backend-specific data. Not sure if this solution is ideal either — there may be something better still. But it seems to have fewer risks than this option, where `SchemaInfo` carries all the backend-specific fields directly, even when they are not allowed to be used.

Reply:
Very much agree.

Btw, the schema info as it was is precisely the `BackendSpecificDataT` you're talking about. Not sure what the right thing to do is, as I don't understand what the goal is very clearly.